src/sDaiProxyActions.sol

```diff
@@ -84,78 +84,71 @@ contract sDaiProxyActions {
         // Joins wad amount to the vat balance
         daiJoin_join(daiJoin, address(this), wad);
         // Approves the adapter to take out DAI from the proxy's balance in the vat
-        if (VatLike(vat).can(address(this), address(savingsJoin)) == 0) {
-            VatLike(vat).hope(savingsJoin);
+        if (vat.can(address(this), address(savingsJoin)) == 0) {
+            vat.hope(savingsJoin);
         }
         // Exits the wad value (equivalent to the DAI wad amount) to Savings Dai
-        SavingsJoinLike(savingsJoin).exit(address(this), wad);
+        SavingsJoinLike(savingsJoin).exit(address(this), mul(wad, ONE) / pot.chi());
     }
 
     function sDaiExit(
         address daiJoin,
         address savingsJoin,
         uint wad
     ) public {
-        address vat = address(SavingsJoinLike(savingsJoin).vat());
-        address pot = address(SavingsJoinLike(savingsJoin).pot());
-        address sDai = address(SavingsJoinLike(savingsJoin).sDai());
+        VatLike vat = SavingsJoinLike(savingsJoin).vat();
+        PotLike pot = SavingsJoinLike(savingsJoin).pot();
+        DSTokenLike sDai = SavingsJoinLike(savingsJoin).sDai();
         // Executes drip to count the savings accumulated until this moment
-        PotLike(pot).drip();
+        pot.drip();
+        // Calculates the pie value in the pot equivalent to the DAI wad amount
+        uint pie = mul(wad, ONE) / pot.chi();
         // Exits DAI from the sDai
-        DSTokenLike(sDai).approve(savingsJoin, wad);
-        // Join Savings Dai back into the Vat
-        SavingsJoinLike(savingsJoin).join(address(this), wad);
+        sDai.approve(savingsJoin, pie);
+        SavingsJoinLike(savingsJoin).join(address(this), pie);
         // Checks the actual balance of DAI in the vat after the pot exit
-        uint pie = mul(PotLike(pot).chi(), wad) / ONE;
-        uint bal = VatLike(vat).dai(address(this));
+        uint bal = vat.dai(address(this));
         // Allows adapter to access to proxy's DAI balance in the vat
-        if (VatLike(vat).can(address(this), address(daiJoin)) == 0) {
-            VatLike(vat).hope(daiJoin);
+        if (vat.can(address(this), daiJoin) == 0) {
+            vat.hope(daiJoin);
         }
         // It is necessary to check if due rounding the exact wad amount can be exited by the adapter.
         // Otherwise it will do the maximum DAI balance in the vat
         DaiJoinLike(daiJoin).exit(
             msg.sender,
-            bal == mul(pie + 1, ONE) ? bal / ONE : pie
+            bal >= mul(wad, ONE) ? wad : bal / ONE
         );
     }
 
-    event log(bytes32, uint);
-
     function sDaiExitAll(
         address daiJoin,
         address savingsJoin
     ) public {
-        address vat = address(SavingsJoinLike(savingsJoin).vat());
-        address pot = address(SavingsJoinLike(savingsJoin).pot());
-        address sDai = address(SavingsJoinLike(savingsJoin).sDai());
+        VatLike vat = SavingsJoinLike(savingsJoin).vat();
```

Comment:
I feel like other of our repos are a bit inconsistent with using `address` or `ContractLike` for types.  Is this is preferred, newer method?

Reply:
In general I like to keep function params and return values with plain `address` type. This way the interoperability between contracts is easier. It also saves some issues when applying inheritance.
Then it will depend on how we use the variables within a function how I use to initiate them. If I need to pass them to other function calls, I use to keep them as `address` and cast them whenever I need to call a method.
However if the contracts are saved in variables just for calling their functions I tend to assigne them directly casted.
Anyway this is my personal behavior about this and we might set some discussion on this topic.